functors task (1)/student/include/impl.h

```diff
@@ -1,48 +1,81 @@
 #pragma once
-#include "bar_serving.h"
+#include <cstdint>
 #include <functional>
+#include <map>
+#include <utility>
+
+#include "bar_serving.h"
 
 /**
  * @todo Implement functor-generator that will return next beer brand (cyclic)
- * @example HoeGaarden -> Corona -> Carlsberg -> Bud -> ZlataPraha -> Leffe -> HoeGaarden...
+ * @example HoeGaarden -> Corona -> Carlsberg -> Bud -> ZlataPraha -> Leffe ->
+ * HoeGaarden...
  */
-struct BeerOrganizer
-{
+struct BeerOrganizer {
+  // we can always be sure that generated beerBrand will be strictly between
+  // None and Max exclusive
+  BeerOrganizer() : m_beer{BeerBrand::None} {}
+
+  BeerBrand operator()() {
+    m_beer = static_cast<BeerBrand>(static_cast<uint>(m_beer) + 1);
+    if (m_beer == BeerBrand::Max) {
+      m_beer = static_cast<BeerBrand>(static_cast<uint>(BeerBrand::None) + 1);
+    }
+    return m_beer;
+  }
+
+ private:
+  BeerBrand m_beer;
 };
 
 /**
- * @todo Implement unary predicate function that will determine if beer is expensive
+ * @todo Implement unary predicate function that will determine if beer is
+ * expensive
  * @return true in case if beer is expensive, false if not
  *
  * @note Only Corona and HoeGaarden are expensive
  */
-bool isExpensiveBeer(/**???*/)
-{
+bool isExpensiveBeer(BeerBrand beer) {
```

Comment:
_don't forget about const :)_

Reply:
Does it is really needed when we passed arguments by value?
I mean, the caller is anyway guaranteed that the original value won't change.

Reply:
Just a style: always mark function params as const, because you can modify it accidentally